applyStyle-related fixes and tests

fleur opened the pull request:
Cleaned up `applyStyle()`, including:
* one of `resolve()` or `reject()` is always called
* returns after `reject()` or `resolve()` is called to ensure rest of function doesn't execute
* refactored  sprite retrieval fetch slightly to improve readability
* all `reject()` or `throw` use Error objects
* improved validation to only allow on and only call for Vector[Tile]Layers

Improved tests:
* added fixtures
* added `applyStyle.test.js` tests
* moved applyStyle test from `test.js` to `applyStyle.test.js`
* mocked out `handles vector sources from TileJSON` test in `test.js`

Miscellaneous:
* added .DS_Store to `.gitignore`
* added `"jest": true` to `__tests__/.eslintrc`

fleur commented on __tests__/test.js:
```diff
@@ -32,21 +33,6 @@ describe('ol-mapbox-style', function() {
     });
   });
 
-  describe('applyStyle', function() {
```
I moved it to the new applyStyle.tests.js for consistency.

ahocevar replied:
Found. Thanks for the clarification.

fleur commented on __tests__/test.js:
```diff
@@ -151,36 +137,29 @@ describe('ol-mapbox-style', function() {
     });
 
     it('handles vector sources from TileJSON', function(done) {
-      var style_url = 'https://rawgit.com/PetersonGIS/CamoStyle/b783aadd625bf0d874f77daa6c597b585f0b63fd/camo3d.json';
```
I added it as a local fixture and modified the test to use that instead of downloading it.  The test is still there.  It didn't seem that the test needed to connect to the outside world.  Does it?

ahocevar replied:
Thanks, I found the test. A local fixture works fine.

fleur commented on __tests__/applyStyle.test.js:
```diff
@@ -0,0 +1,233 @@
+/*eslint no-console: "off"*/
+import 'isomorphic-fetch';
+import nock from 'nock';
+nock.disableNetConnect();
```
It can.  It's just a style preference thing.  In general, I put global modifications to a package right after import so it's obvious something global about the package has been modified.  I'll move it.

fleur commented on index.js:
```diff
@@ -137,27 +152,30 @@ export function applyStyle(layer, glStyle, source, path, resolutions) {
       fetch(spriteUrl, {credentials: 'same-origin'})
         .then(function(response) {
-          // if the response is ready return the JSON promise
-          if (response.status === 200) {
```
It's covered.  It's the equivalent of (status >= 200 && status < 300).

ahocevar replied:
Thanks for the clarification.

fleur commented on index.js:
```diff
@@ -273,6 +296,7 @@ function processStyle(glStyle, map, baseUrl, host, path, accessToken) {
       setBackground(map, glLayer);
     } else {
       id = glLayer.source || getSourceIdByRef(glLayers, glLayer.ref);
+      // TODO: this technique assumes gl layers will be in particular order; fix
```
My use case for this package will be fairly small mapbox styles, and I could easily see just editing them by hand.  I could see hours of pain for someone who didn't know that order mattered.

ahocevar replied:
Can you share an example style where this would fail? Still trying to understand.

ahocevar replied:
I think we can assume validated gl style documents. A validator is available: https://github.com/mapbox/mapbox-gl-js/tree/master/src/style-spec#gl-style-validate (not only as command line utility - it can also be used by an application).

Can this technique also be a problem with valid style documents?

fleur replied:
If there's a validator, then you're right; it probably doesn't matter.  Can I leave the comment in if I take off the "fix" part?  It took me awhile to figure out exactly how that worked, and if anyone else goes in there, it will probably save them some time...

ahocevar replied:
Ok.

fleur commented on index.js:
```diff
@@ -251,12 +269,17 @@ function processStyle(glStyle, map, baseUrl, host, path, accessToken) {
+            layer.setVisible(true);
+          }, function(e) {
+            /*eslint no-console: ["error", { allow: ["error"] }] */
+            console.error(e);
```
Yes, definitely.  `finalizeLayer` needs more work, and I was going to put it in a different PR for some reason I can't remember now.  I'll go ahead and put it in this one.  Stand by.

ahocevar replied:
Fine with me if done in a follow-up pull request.

fleur commented on index.js:
```diff
@@ -111,22 +111,37 @@ function toSpriteUrl(url, path, extension) {
     var spriteScale, spriteData, spriteImageUrl, style;
     function onChange() {
       if (!style && (!glStyle.sprite || spriteData)) {
         if (layer instanceof VectorLayer || layer instanceof VectorTileLayer) {
```
It still needs to be there because int the documentation it says`applyStyle()` can also be called directly, no?

ahocevar replied:
You already added a `reject` in `applyStyle()` when called with a different kind of layer.

For clarification: I meant you can remove the instanceof check from `onChange()`.

fleur replied:
The problem is in the codepath that creates layers for the user.

`processStyle()` is called from `apply()`, and `processStyle()` calls `finalizeLayer()` on _all_ layers the package creates, including geoJSON and raster layers that may have been created.  The check is done because we don't want `applyStyle()` to fail on layers we create for the user.

If this check is not done, it will cause the promise returned from `applyStyle()` in `onChange()` to reject.  I.e. Layers created by the package itself will cause rejection if this check is taken away.  This means that a mapbox style with anything but vector layers can't be passed to `apply()` without the package erroring out.

The check doesn't have to be here.  We could do a check in the monster for loop to not call `finalizeLayer()` except on vector layers.  Would that be better?

ahocevar replied:
Thanks for the explanation. `finalizeLayer()` needs to be called for all layers to make them visible. What about changing things so inside `finalizeLayer()`, `applyStyle()` is only called for `Vector` and `VectorTile` layers?

ahocevar replied:
I've added a commit that removes the redundant check.

ahocevar commented on __tests__/applyStyle.test.js:
```diff
@@ -0,0 +1,237 @@
+    });
+  });
+
+  // TODO: still need to figure out best way to validate sources
```
Does this test actually fail? It would be nice to have it in the test suite.

fleur replied:
The main problem is I don't quite understand the codebase yet.

When `applyStyle()` is called from `apply()`, source ends up being an array with the mapbox id of every layer, so I think what would have to happen is each layer would have to be dereferenced to get it's source type and check it.

I'm hoping a better solution will present itself as I become more familiar with the codebase.  I have a vague idea that the actual source names are being accumulated somewhere and maybe that could be passed instead?

Can I put it off to a future PR, or do I need to dig into it now?  It should probably say "right way", not "best way".  :-)

ahocevar replied:
`apply()` optimizes the gl `layer` to ol `Source` mapping, i.e. it avoids creating multiple vector sources if the gl `layer`s use the same gl `source-layer`. So it can also call `applyStyle` several times with different subsets of the gl layers.

So the test you commented out here makes sense. To make it pass, you'd have to `reject` at the end of the "monster for-loop" as you call it, when `mapboxLayers` has a length of `0`.

ahocevar replied:
Let's handle this in a separate pull request.

ahocevar replied:
Addressed with #109.

ahocevar commented:
> From what I've read, return resolve() is a convention that just makes sure that a function won't execute any more code after it's finished. It's the same as resolve(); return;. The places I don't use return are places where the function would end anyway. I just didn't put it where it didn't need to be.

Makes sense. Thanks!

fleur commented:
Also, I just noticed that the jenkins run failed, possibly because I missed mocking some http call, which is weird because it doesn't fail locally.  I'll need to look into that as well.

ahocevar commented:
> Also, I just noticed that the jenkins run failed, possibly because I missed mocking some http call, which is weird because it doesn't fail locally. I'll need to look into that as well.

You've set `nock` to disallow net requests, this is why the tests fail when fetching tiles or TileJSON metadata. An example error message from Jenkins is
```
Error: NetConnectNotAllowedError: Nock: Disallowed net connect for "api.tiles.mapbox.com:443/v3/mapbox.geography-class.json?secure"
```

fleur commented:
> You've set `nock` to disallow net requests, this is why the tests fail when fetching tiles or TileJSON metadata.

If only it were that simple.  :-)  I wasn't getting the errors locally until I actually added the `nock.disableNetConnect()` to the `tests.js` file, so there's some scoping/caching thing in jenkins I'm unaware of since the test files appear to be called in the same order both locally and in jenkins.
Then when I added it in and started tracking http calls down, tests started behaving differently on separate runs.  

I'm fairly sure it's the tests that call `apply()` and then do the actual tests in a listener, so race condition.  Since the whole point of my digging into this is that very issue, I punted and just removed `nock.disableNetConnect()` entirely to deal with later.

I've just pushed a commit with `finalizeLayer()` redone and am waiting for the jenkins results now.

ahocevar commented:
> I'm fairly sure it's the tests that call apply() and then do the actual tests in a listener, so race condition

It's not a race condition. It's just that depending on how fast the tests complete, tiles may or may not start loading after all metadata has been loaded. And the fixtures do not contain tiles.

ahocevar commented:
Thanks @fleur for this improvement.

ahocevar commented:
@fleur Actually there were several issues with the tests, but I was able to fix them with #109. Thanks for the tests you added, they helped a lot in sanity-checking of the whole async flow.

fleur commented:
@ahocevar Cool, glad I could help.  Unfortunately, I'm now getting transpilation errors running the tests, even when I clone a new repo.  Just doing `npm install && npm test` gets me the following.  Did the babel config change get the wrong version check in maybe?
```
 FAIL  __tests__/test.js
  ● Test suite failed to run

    /Users/fleur/work/ol-mapbox-style/__tests__/test.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,global,jest){import should from 'should';
                                                                                                    ^^^^^^

    SyntaxError: Unexpected identifier

      at ScriptTransformer._transformAndBuildScript (node_modules/jest-runtime/build/script_transformer.js:403:17)
```
Babel also has an api with a call to get the current env, if you're interested:
https://babeljs.io/docs/en/config-files#apienv

ahocevar commented:
I think this is because the `should` import in `__test__/test.js` should be
```js
import should from 'should/as-function';
```
Can you verify that? Strangely, I do not get an error.

ahocevar commented:
I just tried a fresh install, and did not get this error. Our CI suite also did not complain. Wondering what's causing this on your end.

fleur commented:
Thanks for checking.  I dived in further and removing `node_modules` doesn't clear the jest cache.  A `node node_modules/.bin/jest --clearCache` fixed it.  Turns out the default cache directory is user based rather than project based, so because I'm working on multiple projects that use jest, it was picking up wrong versions of stuff.

It would be a good idea to add something like the following to the jest config in `package.json`.  Would have saved me some time.  I googled around a bit and was really surprised I didn't find anything about a best practice configuring the jest cache to be project local.

```
    "cacheDirectory": "<rootDir>/__tests__/jest_cache",
```
